libraries/AP_Generator/AP_Generator_IE_FuelCell.cpp

```diff
@@ -0,0 +1,441 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_Generator_IE_FuelCell.h"
+
+#if GENERATOR_ENABLED
+
+#include <AP_Logger/AP_Logger.h>
+#include <stdio.h>
+
+#define AMS_TO_MAH  1/3600    //Amp.millisec to milliAmp.hour = 1/1E3(ms->s) * 1/3600(s->hr) * 1000(A->mA)
+#define PRESS_GRAD  1/295    //Pressure gradient for for pressure to % conversion
+
+// Constructor
+AP_Generator_IE_FuelCell::AP_Generator_IE_FuelCell(AP_Generator &frontend, AP_Generator::GenType type) :
+    AP_Generator_Backend(frontend),
+    // Reboot required for change in type
+    _type(type)
+{
+}
+
+// Initialize the fuelcell object and prepare it for use
+void AP_Generator_IE_FuelCell::init()
+{
+    _uart = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_Generator, 0);
+
+    if (_uart != nullptr) {
+        _uart->begin(AP::serialmanager().find_baudrate(AP_SerialManager::SerialProtocol_Generator, 0));
+    } else {
+        gcs().send_text(MAV_SEVERITY_INFO, "No serial port found");
+    }
+}
+
+
+// Update fuelcell, expected to be called at 20hz
+void AP_Generator_IE_FuelCell::update()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+
+    uint32_t now = AP_HAL::millis();
+
+   // Read any available data
+    int16_t nbytes = _uart->available();
```

Comment:
That's not the return type from `available()`.

The result from available() should be clamped to, say, 1024 bytes to make sure we don't spend too much time parsing here (it's a precaution we take elsewhere)